### DISCUSSION TITLE
Getting 400 APIErrorCode.CANT_DETERMINE_VALID_WORKSPACE

### DISCUSSION THREAD

**Discussion by shirkan** [fief-dev/fief#145] [Q&A] [answered]
Hi,
I've created a docker compose yml file as described in [https://docs.fief.dev/self-hosting/deployment/docker-compose/](url) using postgres, checking fief server logs i see:

> INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
Main Fief workspace already exists
Main Fief user already exists
INFO:     Started server process [1]
INFO:     Waiting for application startup.
2023-02-21 16:27:41.958 | INFO     | fief.app:on_startup:40 - Fief Server started - {"version": "0.22.1"}
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)

yet when I try to connect on port 8000 I get this in log:

> INFO:     <my_ip>:64434 - "GET / HTTP/1.1" 400 Bad Request

and in the browser I see 

> 400
> APIErrorCode.CANT_DETERMINE_VALID_WORKSPACE

what am i missing?
Thanks!

**Comment by fief-bailiff** [comment]
Hail, @shirkan 👋 Welcome to Fief's kingdom!

Our team will get back to you very soon to help.

In the meantime, take a minute to star our repository ⭐️

![star-fief](https://user-images.githubusercontent.com/1144727/192523552-dc2c31fa-ae94-42fa-9ab4-7d0ba51bb7b0.png)

Farewell!

**Comment by frankie567** [comment, marked as answer]
Hi @shirkan :wave:

This error happens when Fief server was not able to determine the workspace given the hostname.

The environment variable `FIEF_DOMAIN` should match with the hostname you want to use to access the server. Examples:

* `FIEF_DOMAIN=localhost:8000` -> Open http://localhost:8000/admin/
* `FIEF_DOMAIN=fief.mydomain.com` -> Open http://fief.mydomain.com/admin/

**Reply by shirkan** [reply]
Hi @frankie567 thanks for replying!

my .env file looks like this:

```
SECRET=<XXX>
FIEF_CLIENT_ID=<XXX>
FIEF_CLIENT_SECRET=<XXX>
ENCRYPTION_KEY=<XXX>
PORT=8000
ROOT_DOMAIN=localhost:8000
FIEF_DOMAIN=localhost:8000
FIEF_MAIN_USER_EMAIL=<XXX>
FIEF_MAIN_USER_PASSWORD=<XXX>

DATABASE_TYPE=POSTGRESQL
DATABASE_HOST=postgres
DATABASE_PORT=5432
DATABASE_USERNAME=<XXX>
DATABASE_PASSWORD=<XXX>
DATABASE_NAME=fief

REDIS_URL=redis://redis:6379
```

DATABASE_USERNAME, DATABASE_PASSWORD and DATABASE_NAME are equal to POSTGRES_USER, POSTGRES_PASSWORD, POSTGRES_DB respectively in the docker compose yaml of course. 

Indeed i've been trying http://localhost:8000/admin/ but received a 400

**Reply by frankie567** [reply]
Weird! You confirm you get an error like the following:

<img width="1840" alt="Capture d’écran 2023-02-22 à 11 54 04" src="https://user-images.githubusercontent.com/1144727/220599954-637084e3-f2af-4fe8-9c28-7df387ec6e1c.png">

Also, are there any error logs on the server? You can check them with:

```bash
docker-compose logs -f fief-server
```

**Reply by shirkan** [reply]
Yes, this is the screen I'm seeing.
I've attached logs in the main post, attaching again:

```$ sudo docker logs -f shirkython_fief-server_1
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
Main Fief workspace already exists
Main Fief user already exists
INFO:     Started server process [1]
INFO:     Waiting for application startup.
2023-02-21 16:27:41.958 | INFO     | fief.app:on_startup:40 - Fief Server started - {"version": "0.22.1"}
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
INFO:     <MY_IP>:64434 - "GET / HTTP/1.1" 400 Bad Request
INFO:     <MY_IP>:64434 - "GET /admin.css HTTP/1.1" 304 Not Modified
INFO:     <MY_IP>:64434 - "GET /illustrations/broken-soldier.svg HTTP/1.1" 304 Not Modified
INFO:     <MY_IP>:64434 - "GET /illustrations/grass.svg HTTP/1.1" 304 Not Modified
INFO:     <MY_IP>:64434 - "GET /favicon.svg HTTP/1.1" 304 Not Modified
INFO:     <MY_IP>:64437 - "GET / HTTP/1.1" 400 Bad Request
INFO:     <MY_IP>:64438 - "GET /login/ HTTP/1.1" 307 Temporary Redirect
INFO:     <MY_IP>:64438 - "GET /login HTTP/1.1" 400 Bad Request
INFO:     <MY_IP>:64438 - "GET /favicon.ico HTTP/1.1" 404 Not Found
INFO:     <MY_IP>:64451 - "GET /userinfo/ HTTP/1.1" 404 Not Found
INFO:     <MY_IP>:50348 - "GET / HTTP/1.1" 400 Bad Request
INFO:     <MY_IP>:50351 - "GET / HTTP/1.1" 400 Bad Request
WARNING:  Invalid HTTP request received.
INFO:     205.210.31.16:60872 - "GET / HTTP/1.1" 404 Not Found
INFO:     185.180.143.136:35430 - "GET / HTTP/1.1" 404 Not Found
WARNING:  Invalid HTTP request received.
WARNING:  Invalid HTTP request received.
INFO:     167.248.133.61:35332 - "GET / HTTP/1.1" 404 Not Found
INFO:     167.248.133.61:37074 - "GET / HTTP/1.1" 404 Not Found
WARNING:  Invalid HTTP request received.
INFO:     87.236.176.243:48917 - "GET / HTTP/1.1" 404 Not Found
```


unmasked ips are not mine, probably some scanners.

**Reply by frankie567** [reply]
Oh yes, sorry! The lines `WARNING:  Invalid HTTP request received.` bug me a bit.

> unmasked ips are not mine, probably some scanners.

So... You've deployed on a real server that's exposed to the internet? Did you setup some reverse proxy or things like that?

**Reply by shirkan** [reply]
> Oh yes, sorry! The lines `WARNING: Invalid HTTP request received.` bug me a bit.

I don't think it's from my requests, but other scanners.
 
> So... You've deployed on a real server that's exposed to the internet? Did you setup some reverse proxy or things like that?

The server is exposed to the internet but there's no reverse proxy. Even if I use curl internally I get the error:

```
$ curl http://0.0.0.0:8000/admin/


<!DOCTYPE html>
<html lang="en">
  <head>
    
      
        <title id="head-title" hx-swap-oob="true">Error 400</title>
      
      <meta name="viewport" content="width=device-width, initial-scale=1">
      <link rel="icon" href="http://0.0.0.0:8000/admin/static/favicon.svg" />
      <link href="http://0.0.0.0:8000/admin/static/admin.css" rel="stylesheet">
      <script src="https://unpkg.com/htmx.org@1.8.4"></script>
      <script src="https://unpkg.com/hyperscript.org@0.9.7"></script>
      
    
  </head>
  <body class="font-inter antialiased bg-gray-100 text-gray-600">
    
    
<main class="bg-white">

  <div class="fixed bottom-0 right-0 hidden md:block w-1/6 -z-1">
    <img src="http://0.0.0.0:8000/admin/static/illustrations/broken-soldier.svg" alt="Fief Castle" />
  </div>

  <div class="relative flex bg-fixed bg-repeat-x bg-bottom" style="background-image: url('http://0.0.0.0:8000/admin/static/illustrations/grass.svg')">
    <div class="w-full min-h-screen h-full flex justify-center items-center">
      <div class="text-center">
        <h1 class="text-9xl font-bold">400</h1>
        <p class="text-2xl">APIErrorCode.CANT_DETERMINE_VALID_WORKSPACE</p>
      </div>
    </div>
  </div>
</main>

  </body>
</html>
```

**Reply by frankie567** [reply]
Ok so, this server is *different* from the computer you're using?

**Reply by shirkan** [reply]
> Ok so, this server is _different_ from the computer you're using?

No, it's the same one. For now, I'm trying to access it by http://<servers ip>:8000/admin/ 
But as you can see both ways fail.
Here are the running dockers, maybe this could give a hint?

```
$ sudo docker container ls -a
CONTAINER ID   IMAGE                          COMMAND                  CREATED        STATUS        PORTS                                       NAMES
16ab359c12a0   ghcr.io/fief-dev/fief:latest   "fief run-server --p…"   21 hours ago   Up 21 hours   0.0.0.0:8000->8000/tcp, :::8000->8000/tcp   shirkython_fief-server_1
06be70cb1099   postgres:alpine                "docker-entrypoint.s…"   21 hours ago   Up 21 hours   5432/tcp                                    shirkython_postgres_1
19738ba79d0a   redis:alpine                   "docker-entrypoint.s…"   21 hours ago   Up 21 hours   6379/tcp                                    shirkython_redis_1
7d4f20d2c0ff   ghcr.io/fief-dev/fief:latest   "fief run-worker -p …"   21 hours ago   Up 21 hours   8000/tcp                                    shirkython_fief-worker_1
```

**Reply by frankie567** [reply]
Hmm... The key thing is that the hostname in your HTTP request must match the one in DB. So `curl http://localhost:8000/admin/` *should* work, `curl http://:8000/admin/` or `curl http://0.0.0.0:8000/admin/` *won't*.

To be sure we're not missing something obvious here, could you wipe your PostgreSQL database so we're sure we start with clean data? You can run the following command to remove everything, including volumes:

```sh
docker-compose down -v
```

**Reply by shirkan** [reply]
Ok so I changed `FIEF_DOMAIN=0.0.0.0:8000` and now I get a good response for locally running `curl http://0.0.0.0:8000/admin/` - I can see in the response this is a good log in page.
Now I'm trying to set a dns name so I can connect from outside e.g. `fief.mydomain.com`. I'm using nginx to tunnel traffic from port 80 to 443 and add ssl cert for security. The nginx config looks like this (notice I have another server on the same instance, locally listening to port 2323 and tunneled to api.mydomain.com):

```
server {
    listen [::]:80;
    listen 80;
    server_name api.mydomain.com;
    return 301 https://api.mydomain.com$request_uri;
}

server {
    listen [::]:80;
    listen 80;
    server_name fief.mydomain.com;
    return 301 https://fief.mydomain.com$request_uri;
}

server {
     listen [::]:443 ssl http2;
     listen 443 ssl http2;

     server_name api.mydomain.com;

     ssl_certificate /etc/letsencrypt/live/api.mydomain.com/fullchain.pem;
     ssl_certificate_key /etc/letsencrypt/live/api.mydomain.com/privkey.pem;

    location / {
         proxy_pass http://0.0.0.0:2323;
         proxy_redirect off;
         proxy_read_timeout    90;
         proxy_connect_timeout 90;
         proxy_set_header  X-Real-IP  $remote_addr;
         proxy_set_header  X-Forwarded-For $proxy_add_x_forwarded_for;
         proxy_set_header  Host $http_host;
     }
}

server {
     listen [::]:443 ssl http2;
     listen 443 ssl http2;

     server_name fief.mydomain.com;

     ssl_certificate /etc/letsencrypt/live/fief.mydomain.com/fullchain.pem;
     ssl_certificate_key /etc/letsencrypt/live/fief.mydomain.com/privkey.pem;

    location / {
         proxy_pass http://0.0.0.0:8080;
         proxy_redirect off;
         proxy_read_timeout    90;
         proxy_connect_timeout 90;
         proxy_set_header  X-Real-IP  $remote_addr;
         proxy_set_header  X-Forwarded-For $proxy_add_x_forwarded_for;
         proxy_set_header  Host $http_host;
     }
}
```

Does this makes sense?
Do I need to change `FIEF_DOMAIN` to `fief.mydomain.com` ?
Notice I've already created certificates using letsencrypt

Thanks again!

**Reply by shirkan** [reply]
UPDATE: I changed my nginx configuration and now I'm able to access from https://fief.mydomain.com/admin/ endpoint to the dashboard. The problem is it won't load correctly and won't navigate to clients, tenants and so due to some error:
`Failed to load response data: No data found for resoutce with given identifier`


![image](https://user-images.githubusercontent.com/860107/220782745-b4641805-7582-4696-93cd-86c9c40ce7e5.png)


Can you please advise?
Thanks!

**Reply by shirkan** [reply]
UPDATE 2: I managed to solve by changing nginx config with some help in stack overflow, now it looks like it works.
Thanks again :)

**Reply by frankie567** [reply]
Great to hear!

For reference though:  yes, you'll need to change `FIEF_DOMAIN` to `fief.mydomain.com`; otherwise, you'll get the same error again when accessing `https://fief.mydomain.com`.

Be sure also to set the environment variable `FORWARDED_ALLOW_IPS=*` so Fief server can trust the forwarded headers from your proxy.

**Reply by shirkan** [reply]
> Be sure also to set the environment variable `FORWARDED_ALLOW_IPS=*` so Fief server can trust the forwarded headers from your proxy.

What a great tip, you just saved me another question as I was wondering why would the forms generate using http.
Thank you and good luck! :)
